### PR TITLE
C-01: inflation attack fixes

### PR DESCRIPTION
**Pull request by sunnyvempati**
- introduces a `minimumUsdExchangeAmount` & `maxRoundingLoss` variables. (see comments for descriptions).
- adds tests against inflation attacks described in this gist: https://gist.github.com/sipox11/522bcd08115742990a85dca2909f18f2

### CODE REVIEW

File: markets/spot-market/contracts/storage/AsyncOrder.sol [hunk 1]
```diff
@@ -45,6 +49,13 @@ library AsyncOrder {
 
         token.transferFrom(from, address(this), synthAmount);
         asyncOrderData.totalEscrowedSynthShares += sharesAmount;
+
+        // sanity check to ensure the right shares amount is calculated
+        uint acceptableSynthAmount = convertSharesToSynth(asyncOrderData, marketId, sharesAmount) +
+            maxRoundingLoss;
```
Comment (dbeal-eth) [inline markets/spot-market/contracts/storage/AsyncOrder.sol:55]: shouldn't we be subtracting maxRoundingLoss here?

File: markets/spot-market/contracts/storage/AsyncOrder.sol [hunk 2]
```diff
@@ -2,12 +2,15 @@
...
 
 import "@synthetixio/core-modules/contracts/interfaces/ITokenModule.sol";
+import "@synthetixio/core-contracts/contracts/utils/DecimalMath.sol";
 import "../utils/SynthUtil.sol";
 
 /**
  * @title Async order top level data storage
  */
 library AsyncOrder {
+    error InvalidSharesCalculation(uint256 expected, uint256 actual);
```
Comment (dbeal-eth) [inline markets/spot-market/contracts/storage/AsyncOrder.sol:12]: this error name could be more descriptive to an innocent user. for example something like "InsufficientAmountDeposited" or something like that, could perhaps be more helpful

**Review by dbeal-eth — APPROVED** [review]
nice tests

minor notes

File: markets/spot-market/contracts/storage/AsyncOrder.sol (hunk 1 above)
Comment (sunnyvempati) [reply markets/spot-market/contracts/storage/AsyncOrder.sol:55]: no, the `convertSharesToSynth` will never return _more_ synth than the synth amount sent in.  due to division, it loses some micro precision so adding the maxRoundingLoss to the conversion back

File: markets/spot-market/contracts/storage/AsyncOrder.sol (hunk 2 above)
Comment (sunnyvempati) [reply markets/spot-market/contracts/storage/AsyncOrder.sol:12]: makes sense. changing

**Pull request merged**